### PR TITLE
Theano: deconv2d should also shuffle filter_shape

### PR DESCRIPTION
**Pull request by gvtulder**
The `deconv2d` method in the Theano backend applies a dimshuffle to the `kernel` variable, but I think it should also reorder the values in `filter_shape` to match the new `kernel.shape`.

I tried to make a test for this, but that turns out to be difficult. It seems that Theano doesn't use the `filter_shape` most of the time. (I ran into this problem while computing the gradient of the deconvolution on the old GPU backend.)

### CODE REVIEW

**Comment by fchollet** [comment]
> I tried to make a test for this, but that turns out to be difficult.

Can you explain what is the obstacle? Anything else we can do to guarantee that there is an issue and that this patch fixes it?

**Comment by gvtulder** [comment]
The obstacle is that in most cases Theano ignores the `filter_shape` parameter. I can't figure out when it does, so I don't know how to trigger the problem in an isolated case. I have a fairly complex model where it appears, but only on the old GPU backend. There it goes wrong when the gradient of the deconvolution is used to update the bias. There is a reshape there that is somehow based on the `filter_shape`. It's difficult to say how that got there; there is probably a sequence of optimizations that introduces it.

That aside, I'm pretty sure that the current code is not correct. The `filter_shape` should match the shape of the filter tensor passed to `AbstractConv2d_gradInputs`. However, if the first two dimensions of the variable are swapped in `deconv2d`, that should be done to `filter_shape` as well. (Swapping the filters makes sense because the gradInputs Op expects the filters that were used for the forward convolution, so the input and output channels are swapped.) Currently, the dimensions in the filter shape are not swapped, so the `filter_shape` does not match the shape of the tensor. I checked that by printing the `filter_shape` and the actual shape.

As a longer-term solution/indirect proof, I added some shape checks to this Theano PR: Theano/Theano#5267. They check if the given filter matches the `filter_shape`. If those checks are included, the Keras tests produce an error, which disappears again when the filter shape is shuffled correctly.

**Comment by fchollet** [comment]
LGTM

**Comment by eXaz** [comment]
@gvtulder @yaringal I ran into trouble understanding the idea behind the shuffling dimensions when calling `deconv2d`. You stated that

> Swapping the filters makes sense because the gradInputs Op expects the filters that were used for the forward convolution, so the input and output channels are swapped. <

However [this paper (chapter 4.2)](https://arxiv.org/abs/1603.07285) explains transpose convolution (or deconv) in the following way:
> One way to put it is to note that the kernel defines a convolution, but whether it’s a direct convolution or a transposed convolution is determined by how the forward and backward passes are computed. <

Similarly the explanation in [this paper (appendix p. 12)](https://arxiv.org/abs/1602.05110) supports the notion of transposed convolution simply being the gradient of a convolution with respect to its input.

If we expect `theano.tensor.nnet.abstract_conv.AbstractConv2d_gradInputs` to give us exactly this gradient for a given kernel, why would we want to swap input and output channels first?

>[...] the gradInputs Op expects the filters that were used for the forward convolution [...] <

Which is precisely the reason for my confusion why we would hand over different filters. If we look into the implementation of [AbstractConv_gradInputs](https://github.com/Theano/Theano/blob/master/theano/tensor/nnet/abstract_conv.py#L2099) (for tensorflow there are the equivalent lines), we see that the dimensions of the kernels are actually being shuffled in the exact same way.

I would be grateful for any comments or references that could help me clear up my confusion.

**Comment by gvtulder** [comment]
@eXaz: Not sure if this helps, but it's exactly because of that line in `AbstractConv_gradInputs` you linked to that the input and output dimensions should be swapped.

First consider the Theano side. Suppose that you have a forward convolution that goes from `Nin` feature maps to `Nout` feature maps. Theano then expects the kernel shape to be `(Nout, Nin, h, w)` and will produce an output with shape `(..., Nout, ..., ...)`. For the `AbstractConv_gradInputs`, it does the reverse, so it expects an input of shape `(..., Nout, ..., ...)`, the original forward kernel of shape `(Nout, Nin, h, w)` and will return an output of shape `(..., Nin, ..., ...)`. The key part here is that the forward convolution and the gradient expect the same kernel shape.

The Keras deconvolution layer uses a different kernel shape: if we say the deconv input has the shape `(..., Nout, ..., ...)`, you have to give Keras a kernel with shape `(Nin, Nout, h, w)`. This is not what Theano expects, so the `Nin` and `Nout` dimensions need to be swapped.

The sentences you quote seem to refer to the question whether the filters should be flipped, but that is a different problem. This would be the part where you convolve with `kernel` or with `kernel[:, :, ::-1, ::-1]`.

**Comment by eXaz** [comment]
@gvtulder Yes, thank you it did help!

My confusion came from the following setup: an AE would give me weird results when using convolution for the encoder side and deconvolution on the decoder side, while sharing weights with the conv-layer mirrored on the bottleneck. I simply used the same weight-matrix, even though the number of filters and number of inputs are swapped!

So to calculate the forward pass for the deconvolution layers i would have to actually swap those dimensions before passing them to ``AbstractConv_gradInputs``, since the weights are already in the order we will use for the backward pass (which is exactly the order we use for the forward pass in the conv layer).

To my understanding the flipping filters or not are two different operations: convolution and cross-correlation ([see picture to the right](https://en.wikipedia.org/wiki/Cross-correlation))
